libvast/src/layout_version.cpp

```diff
@@ -0,0 +1,74 @@
+#include "vast/layout_version.hpp"
+
+#include "vast/detail/assert.hpp"
+#include "vast/error.hpp"
+#include "vast/io/read.hpp"
+#include "vast/io/write.hpp"
+#include "vast/logger.hpp"
+
+#include <fstream>
+#include <iterator>
+
+#include "caf/expected.hpp"
+
+namespace {
+
+const char* descriptions[]{
+  "invalid",
+  "v0",
+};
+
+static_assert(vast::layout_version{std::size(descriptions)}
+                == vast::layout_version::count,
+              "Mismatch between number of layout versions and descriptions");
+
+} // namespace
+
+namespace vast {
+
+// This must be in namespace `vast` so it can be found by overload resolution.
+static const char* to_string(vast::layout_version v) {
+  return descriptions[static_cast<uint8_t>(v)];
+}
+
+layout_version read_layout_version(const vast::path& dbdir) {
+  if (!exists(dbdir))
+    return layout_version::invalid;
+  auto versionfile = dbdir / "VERSION";
+  auto contents = io::read(versionfile);
+  if (!contents)
+    return layout_version::invalid;
+  if (contents->empty())
+    return layout_version::invalid;
+  std::string_view sv{reinterpret_cast<const char*>(contents->data()),
+                      contents->size()};
+  // Only read until first newline.
+  if (auto first_newline = sv.find('\n'))
+    sv = std::string_view{sv.data(), first_newline};
+  auto begin = std::begin(descriptions);
+  auto end = std::end(descriptions);
+  auto it = std::find(begin, end, sv);
+  if (it == end)
+    return layout_version::invalid;
+  return static_cast<layout_version>(std::distance(begin, it));
```

Comment:
Can you explain to me how exactly this works? Reading an integer from a file should be quite simple, but why are you using the distance between two iterators here?

Reply:
In the current implementation its not an integer but an enum value - it's using the `descriptions` table to stringify when its written, and using the index inside the description table to destringify when its read.